File: Scala/0001.scala
object Solution {
    def twoSum(nums: Array[Int], target: Int): Array[Int] = {
        val numToIndex = scala.collection.mutable.Map[Int, Int]()

        var result: Option[Array[Int]] = None

        for (i <- nums.indices) {
            val complement = target - nums(i)
            numToIndex.get(complement) match {
                case Some(index) =>
                    result = Some(Array(index, i))
                    // No need to break, just continue to the end of the loop
                case None =>
                    numToIndex(nums(i)) = i
            }
        }

        result.getOrElse(throw new IllegalArgumentException("No two sum solution"))
    }
}

File: Elixir/0001.exs
defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
  def two_sum(nums, target) do
    nums
    |> Enum.with_index()
    |> Enum.reduce_while(%{}, fn {num, index}, num_to_index ->
      complement = target - num

      case Map.get(num_to_index, complement) do
        nil ->
          # Add the current number and its index to the map
          {:cont, Map.put(num_to_index, num, index)}

        complement_index ->
          # Return the indices of the complement and the current number
          {:halt, [complement_index, index]}
      end
    end)
  end
end
